precompiles/staking/method_claim_rewards.go

```diff
@@ -0,0 +1,147 @@
+		originIsDelegator = evm.Origin == delegatorAddr
+	)
+
+	// If the delegator is not the origin nor the caller, it's an unauthorized operation.
+	if !callerIsDelegator && !originIsDelegator {
+		return nil, precompiletypes.ErrInvalidAddr{
+			Got:    delegatorAddr.String(),
+			Reason: "unauthorized to withdraw the delegation rewards for delegator",
+		}
+	}
+
+	// Get delegator Cosmos address.
+	delegatorCosmosAddr, err := precompiletypes.GetCosmosAddress(c.bankKeeper, delegatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get validator Cosmos address.
+	validatorCosmosAddr, err := sdk.ValAddressFromBech32(validatorAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Withdraw all the delegation rewards.
+	// The F1 Cosmos distribution scheme implements an all or nothing withdrawal.
+	// The coins could be of multiple denomination, and a mix of ZRC20 and Cosmos coins.
+	coins, err := c.distributionKeeper.WithdrawDelegationRewards(ctx, delegatorCosmosAddr, validatorCosmosAddr)
+	if err != nil {
+		return nil, precompiletypes.ErrUnexpected{
+			When: "WithdrawDelegationRewards",
+			Got:  err.Error(),
+		}
+	}
+
+	// For all the ZRC20 coins withdrawed:
+	// - Check the amount to unlock is valid.
+	// - Burn the Cosmos coins.
+	// - Unlock the ZRC20 coins.
+	for _, coin := range coins {
+		// Filter out invalid coins.
+		if !coin.IsValid() || !coin.Amount.IsPositive() || !precompiletypes.CoinIsZRC20(coin.Denom) {
+			continue
+		}
+
+		// Notice that instead of returning errors we just skip the coin. This is because there might be
+		// more than one ZRC20 coin in the delegation rewards, and we want to unlock as many as possible.
+		// Coins are locked in the bank precompile, so it should be possible to unlock them afterwards.
+		var (
+			zrc20Addr   = common.HexToAddress(strings.TrimPrefix(coin.Denom, config.ZRC20DenomPrefix))
+			zrc20Amount = coin.Amount.BigInt()
+		)
+
+		// Check if bank address has enough ZRC20 balance.
+		// This check is also made inside UnlockZRC20, but repeat it here to avoid burning the coins.
+		if err := c.fungibleKeeper.CheckZRC20Balance(ctx, zrc20Addr, bank.ContractAddress, zrc20Amount); err != nil {
+			continue
+		}
+
+		coinSet := sdk.NewCoins(coin)
+
+		// Send the coins to the fungible module to burn them.
+		if err := c.bankKeeper.SendCoinsFromAccountToModule(ctx, delegatorCosmosAddr, fungibletypes.ModuleName, coinSet); err != nil {
+			continue
+		}
+
+		if err := c.bankKeeper.BurnCoins(ctx, fungibletypes.ModuleName, coinSet); err != nil {
+			continue
+		}
+
+		// Finally, unlock the ZRC20 coins.
+		if err := c.fungibleKeeper.UnlockZRC20(ctx, zrc20Addr, delegatorAddr, bank.ContractAddress, zrc20Amount); err != nil {
+			continue
+		}
+
+		// Emit an event per ZRC20 coin unlocked.
+		// This keeps events as granular and deterministic as possible.
+		if err := c.addClaimRewardsLog(ctx, evm.StateDB, delegatorAddr, zrc20Addr, zrc20Amount); err != nil {
```

Comment:
Can we add the validator address to this log

Reply:
Added in 2b130fdc0f29a4d958d599d633c42f3987a6e7b5